File: 总题库/15.3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        numsLen = len(nums)
        if numsLen < 3:
            return []

        res = []
        for i in range(numsLen):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left = i + 1
            right = numsLen - 1
            while left < right:
                tempSum = nums[i] + nums[left] + nums[right]

                if tempSum == 0:
                    res.append([nums[i], nums[left], nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
                
                elif tempSum < 0:
                    left += 1
                elif tempSum > 0:
                    right -= 1
        return res

File: 总题库/15.3Sum.cpp
// 排序加双指针
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        unordered_set<int> seen;
        vector<vector<int>> res;
        int numsLen = nums.size();
        if (numsLen < 3) return res;

        for (int i = 0; i < numsLen; i++) {
            if (nums[i] > 0) break; // 减枝
            if (i > 0 && nums[i] == nums[i - 1]) continue;// 去重
            int L = i + 1;
            int R = numsLen - 1;
            while (L < R) {
                int sum = nums[i] + nums[L] + nums[R];
                if (sum == 0) {
                    res.emplace_back(vector<int>({nums[i], nums[L], nums[R]}));
                    while (L < R && nums[L] == nums[L + 1]) L++; // 去重；
                    while (L < R && nums[R] == nums[R - 1]) R--; // 去重；
                    L++;
                    R--;
                }
                else if (sum < 0) L++;
                else if (sum > 0) R--;
            }
        }
        return res;
    }
};